pkg/controller/oam/applicationconfiguration/apply.go

```diff
@@ -19,61 +19,211 @@ package applicationconfiguration
-			ref := runtimev1alpha1.TypedReference{
-				APIVersion: wl.Workload.GetAPIVersion(),
-				Kind:       wl.Workload.GetKind(),
-				Name:       wl.Workload.GetName(),
-			}
 			// TODO(rz): Need to find a way to make sure that this is compatible with the trait structure.
-			if err := fieldpath.Pave(t.UnstructuredContent()).SetValue("spec.workloadRef", ref); err != nil {
+			if err := fieldpath.Pave(t.UnstructuredContent()).SetValue("spec.workloadRef", workloadRef); err != nil {
 				return errors.Wrapf(err, errFmtSetWorkloadRef, t.GetName(), wl.Workload.GetName())
 			}
 
 			if err := a.client.Apply(ctx, t, ao...); err != nil {
 				return errors.Wrapf(err, errFmtApplyTrait, t.GetKind(), t.GetName())
 			}
 		}
+
+		for _, s := range wl.Scopes {
+			return a.applyScope(ctx, wl, s)
+		}
+	}
+
+	return a.dereferenceScope(ctx, namespace, status, w)
+}
+
+func (a *workloads) dereferenceScope(ctx context.Context, namespace string, status []v1alpha2.WorkloadStatus, w []Workload) error {
+	for _, st := range status {
+		toBeDeferenced := st.Scopes
+		for _, wl := range w {
+			if (st.Reference.APIVersion == wl.Workload.GetAPIVersion()) &&
+				(st.Reference.Kind == wl.Workload.GetKind()) &&
+				(st.Reference.Name == wl.Workload.GetName()) {
+				toBeDeferenced = findDereferencedScopes(st.Scopes, wl.Scopes)
+			}
+		}
+
+		for _, s := range toBeDeferenced {
+			if err := a.applyScopeRemoval(ctx, namespace, st, s); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func findDereferencedScopes(statusScopes []v1alpha2.WorkloadScope, scopes []unstructured.Unstructured) []v1alpha2.WorkloadScope {
+	toBeDeferenced := []v1alpha2.WorkloadScope{}
+	for _, ss := range statusScopes {
+		found := false
+		for _, s := range scopes {
+			if (s.GetAPIVersion() == ss.Reference.APIVersion) &&
+				(s.GetKind() == ss.Reference.Kind) &&
+				(s.GetName() == ss.Reference.Name) {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			toBeDeferenced = append(toBeDeferenced, ss)
+		}
+	}
+
+	return toBeDeferenced
+}
+
+func (a *workloads) applyScope(ctx context.Context, wl Workload, s unstructured.Unstructured) error {
+	workloadRef := runtimev1alpha1.TypedReference{
+		APIVersion: wl.Workload.GetAPIVersion(),
+		Kind:       wl.Workload.GetKind(),
+		Name:       wl.Workload.GetName(),
+	}
+
+	// Get Scope again to make sure we work with the most up-to date object.
+	scopeObject := unstructured.Unstructured{}
+	scopeObject.SetAPIVersion(s.GetAPIVersion())
+	scopeObject.SetKind(s.GetKind())
+	scopeObjectRef := types.NamespacedName{Namespace: s.GetNamespace(), Name: s.GetName()}
+	if err := a.rawClient.Get(ctx, scopeObjectRef, &scopeObject); err != nil {
+		return errors.Wrapf(err, errFmtApplyScope, s.GetName())
+	}
+
+	refs := []interface{}{}
+	// TODO(asouza): Need to find a way to make sure that this is compatible with the scope structure.
+	if value, err := fieldpath.Pave(scopeObject.UnstructuredContent()).GetValue("status.workloadRefs"); err == nil {
```

Comment:
I suggest we should move workloadRefs to spec of scope object.

Reply:
That is how we had it and @ryanzhang-oss asked to move to status. Can we make a call on this?

Reply:
https://github.com/crossplane/oam-kubernetes-runtime/pull/44#discussion_r443877680

There must be some mis-communication.  I have the same comment as @wonderflow that was marked as "resolved". The workloadRefs should be in the spec.

Reply:
Moved back top spec.